### DISCUSSION TITLE
Prototype interface

### DISCUSSION THREAD

**Discussion by ghost** [simdjson/simdjson#1937] [Show and tell]
I have started to create an interface to `simdjson` which evaluates values by a prototype structure: [simdjson_peval](https://github.com/kla-sch/simdjson_peval#readme)

The basic idea is to create a prototype that follows the structure of the JSON data. The prototype creates a function to parse the JSON data and save it to the specified locations.

This is very similar to RDMS interfaces like ODBC, where the columns are bound to memory locations. Here, however, with structured data. And arrays can also be resolved and stored in containers.

This makes evaluating the data compact and clear. The prototype shows which JSON structure is processed and where the data is stored. This is particularly helpful for maintenance and troubleshooting.

In addition, the error evaluation is simplified with an error container, which saves all errors during the evaluation so that they can be analyzed at the end.

In terms of performance, it is very similar to a handwritten evaluation. The generated evaluation function consists of lambda functions, which are triggered inline and should basically represent what you would write for a hand-written evaluation. However, the error evaluation may cost something.

**Comment by lemire** [comment]
Your work directly relate to two long-standing open issues:

- https://github.com/simdjson/simdjson/issues/484
- https://github.com/simdjson/simdjson/issues/667

I will be checking out your work in the near future.
